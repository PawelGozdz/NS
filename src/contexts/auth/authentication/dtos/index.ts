export * from './refresh.dto';
export * from './signin.dto';
export * from './signup-integration.dto';
export * from './signup.dto';
export * from './tokens-response.dto';
