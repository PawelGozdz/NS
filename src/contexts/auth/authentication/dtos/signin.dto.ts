import { PickType } from '@nestjs/swagger';

import { GlobalDto } from '@libs/common';

export class SignInDto extends PickType(GlobalDto, ['email', 'password']) {}
