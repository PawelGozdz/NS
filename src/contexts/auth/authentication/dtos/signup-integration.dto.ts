import { PickType } from '@nestjs/swagger';

import { GlobalDto } from '@libs/common';

export class SignUpIntegrationDto extends PickType(GlobalDto, ['email']) {}
