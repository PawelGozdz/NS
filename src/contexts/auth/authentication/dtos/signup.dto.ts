import { GlobalDto } from '@libs/common';
import { PickType } from '@nestjs/swagger';

export class SignUpDto extends PickType(GlobalDto, ['email', 'password']) {}
