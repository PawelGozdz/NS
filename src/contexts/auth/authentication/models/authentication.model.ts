import { TableNames } from '@app/core';
import { EntityId, IAuthUserModel } from '@libs/common';

export class AuthUserModel extends IAuthUserModel {
	static tableName = TableNames.AUTH_USERS;
}

export class AuthUser {
	id: string;

	email: string;

	userId: string;

	hash: string;

	hashedRt: string | null;

	lastLogin: Date | null;

	tokenRefreshedAt: Date | null;

	constructor({
		hash,
		hashedRt,
		userId,
		email,
		id,
		tokenRefreshedAt,
		lastLogin,
	}: {
		hash: string;
		hashedRt: string | null;
		userId: string;
		email: string;
		id: string;
		tokenRefreshedAt: Date | null;
		lastLogin: Date | null;
	}) {
		this.id = id;
		this.hash = hash;
		this.email = email;
		this.hashedRt = hashedRt ?? null;
		this.userId = userId;
		this.tokenRefreshedAt = tokenRefreshedAt;
		this.lastLogin = lastLogin;
	}

	static create({
		id,
		userId,
		hash,
		hashedRt,
		email,
		lastLogin,
		tokenRefreshedAt,
	}: {
		id?: string;
		userId: string;
		email: string;
		hash: string;
		hashedRt?: string | null;
		tokenRefreshedAt?: Date | null;
		lastLogin?: Date | null;
	}) {
		return new AuthUser({
			id: id ? EntityId.create(id).value : EntityId.createRandom().value,
			email,
			userId: EntityId.create(userId).value,
			hash,
			hashedRt: hashedRt ?? null,
			lastLogin: lastLogin ?? null,
			tokenRefreshedAt: tokenRefreshedAt ?? null,
		});
	}
}
