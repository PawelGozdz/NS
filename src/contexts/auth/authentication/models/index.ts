export * from './auth-user.fixture';
export * from './authentication.model';
