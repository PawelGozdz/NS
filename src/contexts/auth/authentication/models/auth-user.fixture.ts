import { AppUtils } from '@libs/common';

import { AuthUser } from './authentication.model';

export class AuthUserFixture {
  static create(optional?: {
    id?: string;
    userId?: string;
    email?: string;
    hash?: string;
    hashedRt?: string;
    tokenRefreshedAt?: Date | null;
    lastLogin?: Date | null;
  }) {
    const id = optional?.id ?? AppUtils.getUUID();
    const userId = optional?.userId ?? AppUtils.getUUID();
    const email = optional?.email ?? 'test@email.com';
    const hash = optional?.hash ?? 'hashedPassword';
    const hashedRt = optional?.hashedRt ?? 'hashedRefreshToken';
    const tokenRefreshedAt = optional?.tokenRefreshedAt ?? new Date();
    const lastLogin = optional?.lastLogin ?? new Date();

    return AuthUser.create({
      id,
      userId,
      email,
      hash,
      hashedRt,
      tokenRefreshedAt,
      lastLogin,
    });
  }
}
