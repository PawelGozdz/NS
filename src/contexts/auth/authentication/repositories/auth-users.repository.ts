import { Database, TableNames } from '@app/database';
import { Injectable } from '@nestjs/common';
import dayjs from 'dayjs';
import { AuthUser, AuthUserDao } from '../models';
import { IAuthUsersRepository } from './auth-users-repository.interface';

@Injectable()
export class AuthUsersRepository implements IAuthUsersRepository {
	constructor(private readonly db: Database) {}

	async create(user: AuthUser): Promise<{ id: string }> {
		const now = dayjs().toDate();

		const userDao = await this.db
			.insertInto(TableNames.AUTH_USERS)
			.values({
				id: user.id,
				email: user.email,
				userId: user.userId,
				hash: user.hash,
				hashedRt: user.hashedRt,
				lastLogin: user.lastLogin,
				tokenRefreshedAt: user.tokenRefreshedAt,
				createdAt: now,
				updatedAt: now,
			})
			.returning('id')
			.executeTakeFirstOrThrow();

		return { id: userDao.id };
	}

	async update(user: AuthUser): Promise<void> {
		await this.db.updateTable(TableNames.AUTH_USERS).set(user).where('userId', '=', user.userId).execute();
	}

	async delete(userId: string): Promise<void> {}

	async getByUserId(userId: string): Promise<AuthUser | undefined> {
		const userDao = await this.db.selectFrom(TableNames.AUTH_USERS).selectAll().where('userId', '=', userId).executeTakeFirst();

		if (!userDao) {
			return undefined;
		}

		return this.mapResponse(userDao);
	}
	async getByUserEmail(email: string): Promise<AuthUser | undefined> {
		const userDao = await this.db.selectFrom(TableNames.AUTH_USERS).selectAll().where('email', '=', email).executeTakeFirst();

		if (!userDao) {
			return undefined;
		}

		return this.mapResponse(userDao);
	}

	mapResponse(user: AuthUserDao): AuthUser {
		return AuthUser.create({
			id: user.id,
			email: user.email,
			userId: user.userId,
			hash: user.hash,
			hashedRt: user.hashedRt,
			lastLogin: user.lastLogin,
			tokenRefreshedAt: user.tokenRefreshedAt,
		});
	}
}
