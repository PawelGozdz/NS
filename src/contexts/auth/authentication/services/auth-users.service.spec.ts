import { CannotCreateUserError, CreateUserIntegrationEvent, GetUserByEmailIntegrationEvent, GetUserByIdIntegrationEvent } from '@app/core';
import { createMock } from '@golevelup/ts-jest';
import { TestLoggerModule, catchActError } from '@libs/testing';
import { EventEmitter2 } from '@nestjs/event-emitter';
import { Test, TestingModule } from '@nestjs/testing';

import { SignUpIntegrationDto } from '../dtos';
import { AuthUserFixture } from '../models';
import { IAuthUsersRepository } from '../repositories';
import { AuthUsersService } from './auth-users.service';

describe('AuthService', () => {
	let service: AuthUsersService;
	let authUsersRepositoryMock: jest.Mocked<IAuthUsersRepository>;
	let eventEmitterMock: jest.Mocked<EventEmitter2>;

	beforeEach(async () => {
		authUsersRepositoryMock = createMock();
		eventEmitterMock = createMock();

		const module: TestingModule = await Test.createTestingModule({
			imports: [TestLoggerModule.forRoot()],
			providers: [
				AuthUsersService,
				{
					provide: IAuthUsersRepository,
					useValue: authUsersRepositoryMock,
				},
				{
					provide: EventEmitter2,
					useValue: eventEmitterMock,
				},
			],
		}).compile();

		service = module.get<AuthUsersService>(AuthUsersService);
	});

	afterEach(() => {
		jest.clearAllMocks();
		jest.resetModules();
		jest.restoreAllMocks();
	});

	const authUser = AuthUserFixture.create();
	const user = {
		email: authUser.email,
		id: authUser.userId,
	};

	const userResp = {
		id: user.id,
	};

	const expectTokens = {
		lastLogin: expect.any(Date),
		tokenRefreshedAt: expect.any(Date),
	};

	describe('create', () => {
		it('should create user', async () => {
			// Arrange
			authUsersRepositoryMock.create.mockResolvedValue(userResp);

			// Act
			const result = await service.create(authUser);

			// Assert
			expect(result).toEqual(userResp);
			expect(result).toMatchSnapshot({
				id: expect.any(String),
			});
		});
	});

	describe('update', () => {
		it('should update user', async () => {
			// Arrange
			authUsersRepositoryMock.update.mockResolvedValue(void 0);

			// Act
			const result = await service.update(authUser);

			// Assert
			expect(result).toBeUndefined();
			expect(authUsersRepositoryMock.update).toHaveBeenNthCalledWith(1, authUser);
		});
	});

	describe('delete', () => {
		it('should delete user', async () => {
			// Arrange
			authUsersRepositoryMock.delete.mockResolvedValue(void 0);

			// Act
			const result = await service.delete(authUser.userId);

			// Assert
			expect(result).toBeUndefined();
			expect(authUsersRepositoryMock.delete).toHaveBeenNthCalledWith(1, authUser.userId);
		});
	});

	describe('getByUserId', () => {
		it('should return user with getByUserId', async () => {
			// Arrange
			authUsersRepositoryMock.getByUserId.mockResolvedValue(authUser);

			// Act
			const result = await service.getByUserId(authUser.userId);

			// Assert
			expect(result).toEqual(authUser);
			expect(authUsersRepositoryMock.getByUserId).toHaveBeenNthCalledWith(1, authUser.userId);
		});
	});

	describe('getByUserEmail', () => {
		it('should return user with getByUserEmail', async () => {
			// Arrange
			authUsersRepositoryMock.getByUserEmail.mockResolvedValue(authUser);

			// Act
			const result = await service.getByUserEmail(authUser.email);

			// Assert
			expect(result).toEqual(authUser);
			expect(authUsersRepositoryMock.getByUserEmail).toHaveBeenNthCalledWith(1, authUser.email);
		});
	});

	describe('getIntegrationUserById', () => {
		it('should return user with getIntegrationUserById', async () => {
			// Arrange
			eventEmitterMock.emitAsync.mockResolvedValue([user]);
			const paramEvent = new GetUserByIdIntegrationEvent(authUser.userId);
			paramEvent.integrationEventId = expect.any(String);
			paramEvent.integrationEventOccuredON = expect.any(Date);

			// Act
			const result = await service.getIntegrationUserById(authUser.userId);

			// Assert
			expect(result).toEqual(user);
			expect(eventEmitterMock.emitAsync).toHaveBeenNthCalledWith(1, GetUserByIdIntegrationEvent.eventName, paramEvent);
		});
	});

	describe('getIntegrationUserByEmail', () => {
		it('should return user with getIntegrationUserByEmail', async () => {
			// Arrange
			eventEmitterMock.emitAsync.mockResolvedValue([user]);
			const paramEvent = new GetUserByEmailIntegrationEvent(authUser.email);
			paramEvent.integrationEventId = expect.any(String);

			// Act
			const result = await service.getIntegrationUserByEmail(authUser.email);

			// Assert
			expect(result).toEqual(user);
			expect(eventEmitterMock.emitAsync).toHaveBeenNthCalledWith(1, GetUserByEmailIntegrationEvent.eventName, paramEvent);
		});
	});

	describe('createIntegrationUser', () => {
		const dto: SignUpIntegrationDto = {
			email: 'test@example.com',
		};

		describe('success', () => {
			it('should return user created id', async () => {
				// Arrange
				eventEmitterMock.emitAsync.mockResolvedValue([userResp]);
				const paramEvent = new CreateUserIntegrationEvent({
					email: dto.email,
					...expectTokens,
				});
				paramEvent.integrationEventId = expect.any(String);

				// Act
				const result = await service.createIntegrationUser(dto);

				// Assert
				expect(result).toEqual(userResp);
				expect(eventEmitterMock.emitAsync).toHaveBeenNthCalledWith(1, CreateUserIntegrationEvent.eventName, paramEvent);
			});
		});

		describe('failure', () => {
			it('should throw CannotCreateUserError if no user returned', async () => {
				// Arrange
				eventEmitterMock.emitAsync.mockRejectedValueOnce(CannotCreateUserError.failed());
				const paramEvent = new CreateUserIntegrationEvent({
					email: dto.email,
					...expectTokens,
				});
				paramEvent.integrationEventId = expect.any(String);

				// Act
				const { error } = await catchActError(() => service.createIntegrationUser(dto));

				// Assert
				expect(error).toBeInstanceOf(CannotCreateUserError);
				expect(eventEmitterMock.emitAsync).toHaveBeenNthCalledWith(1, CreateUserIntegrationEvent.eventName, paramEvent);
			});
		});
	});
});
