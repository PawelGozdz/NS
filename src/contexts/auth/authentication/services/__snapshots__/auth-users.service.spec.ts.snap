// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthService create should create user 1`] = `
{
  "id": Any<String>,
}
`;
